### ISSUE TITLE
Exported routine not executed if last line in sunk pointy block

### ISSUE THREAD

**Issue by landyacht** [rakudo/rakudo#5671]
## The Problem
Consider the following project structure:

**A.raku**
```raku
use lib '.';
use B;
use C;

print-var;
```

**B.rakumod**
```raku
unit module B;

my %handlers;

multi trait_mod:<is>(Sub $sub, :@handler [Str $for]) is export {
    %handlers{$for} = $sub;
}

sub print-var is export {
    say %handlers.gist;
}
```

**C.rakumod**
```raku
use lib '.'; # so this can be executed directly
use B;

unit module C;

sub foo is handler['foo'] {
    say 'foo handler';
}

print-var;
```
 
## Expected Behavior
When `raku A.raku` is executed, the `%handlers` hash is printed twice and has one key both times.

## Actual Behavior
When `raku A.raku` is executed, the `%handlers` hash is printed twice but is empty BOTH times.
However, when `raku C.rakumod` is executed, the hash prints once and does show as containing the expected value.

I suspect this issue has something to do with the fact that `is` logic is compile-time while the printing happens at run-time. It's odd that it works as expected when the handler sub is declared in the same module that it is printed, but the very same line of code behaves differently if that module is pulled in via A.raku rather than executed alone/directly.

## Environment
* Operating system: Gentoo Linux
* Compiler version (`rakudo -v` or `raku -v`): Release 2024.09

**Comment by landyacht** [comment]
Apologies about the totally unrelated previous title. I was working on another issue before changing gears to this :slightly_smiling_face:

**Comment by librasteve** [comment]
When a module is use'd, any actual code lines in the module are executed (ie code that is not just a declaration).

So the explanation for the two print-var outputs is that print-var is called both in A.raku and in C.rakumod.

**Comment by landyacht** [comment]
Just to be clear, it's not unexpected for print-var to run twice, just that it prints an empty hash both times (or that is prints an empty hash at all! The hash ought to have a value in it once runtime starts)

**Comment by landyacht** [comment]
@lizmat why does it remain populated in the case where C.rakumod is executed directly?

Also, if the following phasers are added to B.rakumod, the output of print_var shows the hash as populated:
```raku
BEGIN {
        say "BEGIN %handlers.gist()";
}
CHECK {
        say "CHECK %handlers.gist()";
}
INIT {
        say "INIT  %handlers.gist()";
}
END {
        say "END   %handlers.gist()";
}```

Output:
```
INIT  {}
INIT  {}
END   {foo => &foo}
{foo => &foo}
{foo => &foo}
END   {foo => &foo}```

If the hash being reset between compile- and run-time is intended behavior, do you have any suggestion for a way to register subroutines with a particular trait in a structure that can be accessed at runtime?

**Comment by ab5tract** [comment]
@landyacht Does adding 'no precompilation' to the module that defines the trait_mod change anything?

Also, I noticed that there wasn't any BEGIN appearing in the outputs you shared, though from the snippet you shared I would expect BEGIN to appear.

One last item to note: you don't have to worry about adding .gist in conjunction with 'say' as it will call that method one any variable that is fed to it.

**Comment by landyacht** [comment]
@ab5tract I was also a bit confused that BEGIN and CHECK did not appear in the output. I can imagine that at BEGIN time, the variable might not be initialized or "ready" yet, but in that case I would expect some kind of error, rather than no output at all. Surely at CHECK time, which is defined as "as late as possible" during compile-time, the variable should exist and be populated, though.  I don't know enough about Rakudo/MoarVM internals to hazard a guess at why this is behaving strangely, but I wouldn't be surprised if the original issue is related somehow.

**Comment by ugexe** [comment]
I can't reproduce from the original example:
```
$ raku A.raku
===SORRY!=== Error while compiling /Users/nlogan/repos/asdf/A.raku
===SORRY!=== Error while compiling /Users/nlogan/repos/asdf/C.rakumod (C)
'use lib' cannot be pre-compiled and thus cannot be used in a module
at /Users/nlogan/repos/asdf/C.rakumod (C):1
------> use lib '.'⏏; # so this can be executed directly

at /Users/nlogan/repos/asdf/A.raku:3
```
```
$ raku -v
Welcome to Rakudo™ v2024.09.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2024.09.
```

---

> why does it remain populated in the case where C.rakumod is executed directly?

If you are running `raku C.rakumod` then precompiled code would not be used. To do that you'd have to  `raku -I. -e 'use C'`, but then I'd expect you get the same error I posted above.

**Comment by landyacht** [comment]
@ugexe yes, that's my bad - I added the `use lib` after and didn't re-test :(

What do you think is the "correct" behavior? To me it seems odd to have such easy hooks into compile-time like traits and phasers, yet not be able to carry forward values into runtime this way. At the very least, it seems wrong for this to be inconsistent depending on whether precompilation is involved or not.

Considering the following sample from Synopsis 04, it seems that the ability to carry forward values is intended. Granted, in this example we could imagine the BEGIN portion being replaced with a constant prior to runtime, which is a little different than recording to a hashmap, but the spirit is the same IMO.
```raku
my $compiletime = BEGIN { now };
```
